executor/aggfuncs/func_max_min.go

```diff
@@ -14,121 +14,133 @@
 package aggfuncs
 
 import (
-	"container/heap"
 	"unsafe"
 
+	"github.com/pingcap/errors"
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/stringutil"
 )
 
-type maxMinHeap struct {
-	data    []interface{}
-	varSet  map[interface{}]int64
-	isMax   bool
-	cmpFunc func(i, j interface{}) int
+// newDeque inits a new deque
+func newDeque(isMax bool, cmpFunc func(i, j interface{}) int) *deque {
+	return &deque{[]pair{}, isMax, cmpFunc}
 }
 
-func newMaxMinHeap(isMax bool, cmpFunc func(i, j interface{}) int) *maxMinHeap {
-	h := &maxMinHeap{
-		data:    make([]interface{}, 0),
-		varSet:  make(map[interface{}]int64),
-		isMax:   isMax,
-		cmpFunc: cmpFunc,
-	}
-	return h
+// deque is an array based double end queue
+type deque struct {
```

Comment:
Can you add unit test to ensure the `deque` functionality?

Reply:
Done.